packages/devtools_app/lib/src/service/service_manager.dart

```diff
@@ -180,22 +175,82 @@ class ServiceConnectionManager {
     await serviceManager.isolateManager.init(isolates);
   }
 
+  /// Returns the package root URI for the connected app.
+  ///
+  /// This should be the directory up the tree from the debugging target that
+  /// contains the .dart_tool/package_config.json file.
+  ///
+  /// This method contains special logic for detecting the package root for
+  /// test targets (i.e. a VM service connections spawned from `dart test` or
+  /// `flutter test`). This is because the main isolate for test targets is
+  /// running the test runner, and not the test library itself, so we have to do
+  /// some extra work to find the package root of the test target.
+  Future<Uri?> connectedAppPackageRoot() async {
+    var packageRootUriString = await rootPackageDirectoryForMainIsolate();
+    _log.fine(
+      '[connectedAppPackageRoot] root package directory for main isolate: '
+      '$packageRootUriString',
+    );
+
+    // If a Dart library URI was returned, this may be a test target (i.e. a
+    // VM service connection spawned from `dart test` or `flutter test`).
+    if (packageRootUriString?.endsWith('.dart') ?? false) {
+      final rootLibrary = await _mainIsolateRootLibrary();
+      final testTargetFileUriString =
+          (rootLibrary?.dependencies ?? <LibraryDependency>[])
+              .firstWhereOrNull((dep) => dep.prefix == 'test')
+              ?.target
+              ?.uri;
+      if (testTargetFileUriString != null) {
+        _log.fine(
+          '[connectedAppPackageRoot] detected test library from root library '
+          'imports: $testTargetFileUriString',
+        );
+        packageRootUriString = await packageRootFromFileUriString(
+          testTargetFileUriString,
+          dtd: dtdManager.connection.value,
+        );
+        _log.fine(
+          '[connectedAppPackageRoot] package root for test target: '
+          '$packageRootUriString',
+        );
+      }
+    }
+
+    return packageRootUriString == null
+        ? null
+        : Uri.parse(packageRootUriString);
+  }
+
+  Future<Library?> _mainIsolateRootLibrary() async {
+    final ref = (await serviceManager.isolateManager.waitForMainIsolateState())
+        ?.isolateNow
+        ?.rootLib;
+    if (ref == null) return null;
+    final library = await serviceManager.service!.getObject(
+      serviceManager.isolateManager.mainIsolate.value!.id!,
+      ref.id!,
+    );
+    return library is Library ? library : null;
```

Comment:
Should we `assert(library is Library)` and then `return library as Library` instead? This should never resolve to anything but a `Library`.

Reply:
couldn't this technically be a Sentinel?

Reply:
added a try / catch in case a SentinelException is thrown. PTAL.

Reply:
Yep, that's the right thing to do. LGTM.